diff -ur vice-2.1-orig/src/arch/riscos/uiconfig.c vice-2.1-ef41/src/arch/riscos/uiconfig.c
--- vice-2.1-orig/src/arch/riscos/uiconfig.c	2008-10-10 21:08:26.000000000 +0200
+++ vice-2.1-ef41/src/arch/riscos/uiconfig.c	2009-06-03 19:16:05.000000000 +0200
@@ -1238,7 +1238,7 @@
    CARTRIDGE_FINAL_III, CARTRIDGE_SIMONS_BASIC, CARTRIDGE_OCEAN, CARTRIDGE_EXPERT,
    CARTRIDGE_FUNPLAY, CARTRIDGE_SUPER_GAMES, CARTRIDGE_ATOMIC_POWER, CARTRIDGE_EPYX_FASTLOAD,
    CARTRIDGE_WESTERMANN, CARTRIDGE_REX, CARTRIDGE_FINAL_I, CARTRIDGE_MAGIC_FORMEL,
-   CARTRIDGE_GS, CARTRIDGE_WARPSPEED, CARTRIDGE_DINAMIC, CARTRIDGE_ZAXXON}
+   CARTRIDGE_GS, CARTRIDGE_WARPSPEED, CARTRIDGE_DINAMIC, CARTRIDGE_ZAXXON, CARTRIDGE_EASYFLASH}
 };
 
 static struct MenuDisplayPALDepth {
diff -ur vice-2.1-orig/src/c64/cart/c64cart.c vice-2.1-ef41/src/c64/cart/c64cart.c
--- vice-2.1-orig/src/c64/cart/c64cart.c	2008-11-24 19:42:52.000000000 +0100
+++ vice-2.1-ef41/src/c64/cart/c64cart.c	2009-06-03 19:16:05.000000000 +0200
@@ -309,7 +309,7 @@
     }
 
     /* allocate temporary array */
-    rawcart = lib_malloc(0x88000);
+    rawcart = lib_malloc(0x100000);
 
 /*  cart should always be detached. there is no reason for doing fancy checks
     here, and it will cause problems incase a cart MUST be detached before
diff -ur vice-2.1-orig/src/c64/cart/c64cartmem.c vice-2.1-ef41/src/c64/cart/c64cartmem.c
--- vice-2.1-orig/src/c64/cart/c64cartmem.c	2008-11-21 19:29:58.000000000 +0100
+++ vice-2.1-ef41/src/c64/cart/c64cartmem.c	2009-06-06 12:43:55.000000000 +0200
@@ -79,7 +79,7 @@
 export_t export;
 
 /* Expansion port ROML/ROMH images.  */
-BYTE roml_banks[0x80000], romh_banks[0x20000];
+BYTE roml_banks[0x80000], romh_banks[0x80000];
 
 /* Expansion port RAM images.  */
 BYTE export_ram0[C64CART_RAM_LIMIT];
@@ -94,6 +94,15 @@
 /* Type of the cartridge attached.  */
 int mem_cartridge_type = CARTRIDGE_NONE;
 
+/* EasyFlash GAME/EXROM "Jumper" (in .crt file) configuration */
+BYTE easyflash_game_exrom_jumper;
+
+/* EasyFlash debug */
+BYTE easyflash_debug = 1;
+static const char *easyflash_memconfig[] = {
+	"8k", "16k", "RAM", "ult"
+};
+
 /*
   mode & 3 = 0 : roml
   mode & 3 = 1 : roml & romh
@@ -242,6 +251,41 @@
         cart_ultimax_phi1 = 0;
         cart_ultimax_phi2 = 0;
         break;
+      case CARTRIDGE_EASYFLASH:
+      {
+      	// backup current bank, cause cartridge_config_changed will overwrite it
+      	int easyflash_bank = roml_bank; // roml_bank is always identical to romh_bank
+		if((addr & 1) == 0){
+			// $de00
+			easyflash_bank = value & 0x3f;
+			if(easyflash_debug)
+		      	log_message(LOG_DEFAULT, "EasyFlash: changing bank to %d (addr $%04x, value $%02x)", value & 0x3f, addr, value);
+		}else{
+			// $de01
+			if(value & 0x20){
+				// overwrite jumpers
+		        cartridge_config_changed((value>>6)^1, (value>>6)^1, CMODE_READ);
+				if(easyflash_debug)
+			      	log_message(LOG_DEFAULT, "EasyFlash: changing mode to %s (addr $%04x, value $%02x)", easyflash_memconfig[((value>>6)^1)&3], addr, value);
+		    }else{
+		    	// jumper config
+		        cartridge_config_changed((easyflash_game_exrom_jumper>>6)^1, (easyflash_game_exrom_jumper>>6)^1, CMODE_READ);
+				if(easyflash_debug)
+			      	log_message(LOG_DEFAULT, "EasyFlash: changing mode to jumper %s (addr $%04x, value $%02x)", easyflash_memconfig[((easyflash_game_exrom_jumper>>6)^1)&3], addr, value);
+		    }
+		    if((value & 0x01) && !easyflash_debug){
+		    	easyflash_debug = 1;
+		      	log_message(LOG_DEFAULT, "EasyFlash: enabling debug mode");
+		    }else if((value & 0x02) && easyflash_debug){
+		    	easyflash_debug = 0;
+		      	log_message(LOG_DEFAULT, "EasyFlash: disabling debug mode");
+		    }
+	    }
+        cartridge_romhbank_set(easyflash_bank);
+   	    cartridge_romlbank_set(easyflash_bank);
+        mem_pla_config_changed();
+      }
+        break;
       case CARTRIDGE_GS:
         cartridge_romlbank_set(addr & 0x3f);
         export.game = 0;
@@ -460,6 +504,9 @@
       case CARTRIDGE_MAGIC_FORMEL:
         magicformel_roml_store(addr, value);
         return;
+      case CARTRIDGE_EASYFLASH:
+      	mem_store_without_romlh(addr, value);
+      	return;
     }
 
     if (export_ram)
@@ -496,6 +543,9 @@
       case CARTRIDGE_MAGIC_FORMEL:
         magicformel_romh_store(addr, value);
         return;
+      case CARTRIDGE_EASYFLASH:
+      	mem_store_without_romlh(addr, value);
+      	return;
     }
     if (isepic_enabled && isepic_switch)
     {
@@ -716,6 +766,10 @@
         cartridge_config_changed(1, 1, CMODE_READ);
         cartridge_store_io1((WORD)0xde00, 0);
         break;
+      case CARTRIDGE_EASYFLASH:
+        cartridge_store_io1((WORD)0xde00, 0);
+        cartridge_store_io1((WORD)0xde01, 0);
+        break;
       case CARTRIDGE_GS:
         cartridge_config_changed(0, 0, CMODE_READ);
         cartridge_store_io1((WORD)0xde00, 0);
@@ -849,6 +903,11 @@
         /* Hack: using 16kB configuration, but some carts are 8kB only */
         cartridge_config_changed(1, 1, CMODE_READ);
         break;
+      case CARTRIDGE_EASYFLASH:
+        memcpy(roml_banks, rawcart, 0x2000 * 64);
+        memcpy(romh_banks, &rawcart[0x2000 * 64], 0x2000 * 64);
+        cartridge_config_changed((easyflash_game_exrom_jumper>>6)^1, (easyflash_game_exrom_jumper>>6)^1, CMODE_READ);
+        break;
       case CARTRIDGE_ULTIMAX:
         generic_ultimax_config_setup(rawcart);
         break;
diff -ur vice-2.1-orig/src/c64/cart/c64cartmem.h vice-2.1-ef41/src/c64/cart/c64cartmem.h
--- vice-2.1-orig/src/c64/cart/c64cartmem.h	2008-07-07 20:20:43.000000000 +0200
+++ vice-2.1-ef41/src/c64/cart/c64cartmem.h	2009-06-03 19:31:39.000000000 +0200
@@ -40,5 +40,7 @@
 
 extern BYTE export_ram0[];
 
+extern BYTE easyflash_game_exrom_jumper;
+
 #endif
 
diff -ur vice-2.1-orig/src/c64/cart/crt.c vice-2.1-ef41/src/c64/cart/crt.c
--- vice-2.1-orig/src/c64/cart/crt.c	2008-07-07 20:20:43.000000000 +0200
+++ vice-2.1-ef41/src/c64/cart/crt.c	2009-06-03 19:28:07.000000000 +0200
@@ -190,6 +190,27 @@
             }
         }
         break;
+      case CARTRIDGE_EASYFLASH:
+        easyflash_game_exrom_jumper =
+        	(header[0x18] ? 0x80 : 0) | // EXROM in .crt config
+        	(header[0x19] ? 0x40 : 0); // GAME in .crt config
+        	
+        while (1) {
+            if (fread(chipheader, 0x10, 1, fd) < 1) {
+                fclose(fd);
+                break;
+            }
+            if (chipheader[0xb] >= 64 || (chipheader[0xc] != 0x80
+                && chipheader[0xc] != 0xa0)) {
+                fclose(fd);
+                return -1;
+            }
+            if (fread(&rawcart[(chipheader[0xb] << 13) | (chipheader[0xc] == 0x80 ? 0 : 1<<19)], 0x2000, 1, fd) < 1) {
+                fclose(fd);
+                return -1;
+            }
+        }
+        break;
       case CARTRIDGE_FUNPLAY:
         while (1) {
             if (fread(chipheader, 0x10, 1, fd) < 1) {
diff -ur vice-2.1-orig/src/cartridge.h vice-2.1-ef41/src/cartridge.h
--- vice-2.1-orig/src/cartridge.h	2008-07-07 20:23:21.000000000 +0200
+++ vice-2.1-ef41/src/cartridge.h	2009-06-03 19:16:05.000000000 +0200
@@ -92,6 +92,7 @@
    to become fc2 */
 #define CARTRIDGE_ACTION_REPLAY4     30
 #define CARTRIDGE_STARDOS            31
+#define CARTRIDGE_EASYFLASH          41
 
 /* Expert cartridge has three modes: */
 #define CARTRIDGE_MODE_OFF                      0
